core/arch/x86/mangle.c

```diff
@@ -388,23 +420,39 @@ insert_push_all_registers(dcontext_t *dcontext, clean_call_info_t *cci,
                         opnd_create_reg(REG_SAVED_XMM0 + (reg_id_t)i)));
             }
         }
-        ASSERT(i * MCXT_SIMD_SLOT_SIZE == MCXT_TOTAL_SIMD_SLOTS_SIZE ||
-               i * MCXT_SIMD_SLOT_SIZE == MCXT_TOTAL_SIMD_SSE_AVX_SLOTS_SIZE);
-    }
-    /* TODO i#1312: the zmm and mask fields need to be copied. */
-    /* pc and aflags */
-    if (!cci->skip_save_flags) {
-        ASSERT(offs_beyond_xmm == 0);
-        if (opnd_is_immed_int(push_pc))
-            PRE(ilist, instr, INSTR_CREATE_push_imm(dcontext, push_pc));
-        else
-            PRE(ilist, instr, INSTR_CREATE_push(dcontext, push_pc));
-        dstack_offs += XSP_SZ;
-        PRE(ilist, instr, INSTR_CREATE_pushf(dcontext));
-        dstack_offs += XSP_SZ;
-    } else {
-        ASSERT(offs_beyond_xmm == 2 * XSP_SZ || !cci->preserve_mcontext);
-        /* for cci->preserve_mcontext we added to the lea above so we ignore push_pc */
+        if (ZMM_ENABLED()) {
+            PRE(ilist, instr, INSTR_CREATE_jmp(dcontext, opnd_create_instr(post_push)));
+            PRE(ilist, instr, pre_avx512_push /*label*/);
+            uint opcode_avx512 = move_mm_avx512_reg_opcode(ALIGNED(alignment, 64));
+            for (i = 0; i < proc_num_simd_registers(); i++) {
+                if (!cci->simd_skip[i]) {
+                    instr_t *simdmov = instr_create_1dst_2src(
+                        dcontext, opcode_avx512,
+                        opnd_create_base_disp(REG_XSP, REG_NULL, 0,
+                                              PRE_XMM_PADDING + i * MCXT_SIMD_SLOT_SIZE +
+                                                  offs_beyond_xmm,
+                                              OPSZ_SAVED_ZMM),
+                        opnd_create_reg(DR_REG_K0),
+                        opnd_create_reg(DR_REG_START_ZMM + (reg_id_t)i));
+                    PRE(ilist, instr, simdmov);
+                }
+            }
+            for (i = 0; i < proc_num_opmask_registers(); i++) {
+                if (!cci->opmask_skip[i]) {
+                    instr_t *maskmov = instr_create_1dst_1src(
+                        dcontext,
+                        proc_has_feature(FEATURE_AVX512BW) ? OP_kmovq : OP_kmovw,
+                        opnd_create_base_disp(REG_XSP, REG_NULL, 0,
+                                              PRE_XMM_PADDING +
+                                                  MCXT_TOTAL_SIMD_SLOTS_SIZE +
+                                                  offs_beyond_xmm,
+                                              OPSZ_SAVED_OPMASK),
+                        opnd_create_reg(DR_REG_START_OPMASK + (reg_id_t)i));
+                    PRE(ilist, instr, maskmov);
```

Comment:
There is no computation that needs to be updated as far as I could see. There are no inline restrictions that I put in place in addition to the ones that are already there. The most common one we're running into is the "no PIC" code. Actually not 100% true, there is a small addition when looking at spill registers, it bails with >= 3 mask registers, analog to what was already there for SIMD registers. Are you worried about too much code getting inlined? It would be possible to adjust MAX_NUM_INLINE_INSTRS in this case to account for the bigger meta block. Resolving this comment for now, re-open if necessary.

Reply:
Mostly just looking for a manual sanity check that all this code is out-of-line in the helpers, not inlined.  32-bit may have often had inlined state swaps before?  Is that typically out-of-line now with the 512 code?

Reply:
The last comment was updated/edited.